python/geomGeant4.py

```diff
@@ -148,21 +148,62 @@ def addVMCFields(shipGeo, controlFile = '', verbose = False, withVirtualMC = Tru
 
     # Set the main spectrometer field map as a global field
     if hasattr(shipGeo, 'Bfield'):
+      fieldsList = []
       fieldMaker.defineFieldMap('MainSpecMap', 'files/MainSpectrometerField.root',
                                 ROOT.TVector3(0.0, 0.0, shipGeo.Bfield.z))      
+      fieldsList.append('MainSpecMap')   
       withConstFieldNuTauDet = False
       if hasattr(shipGeo.EmuMagnet,'WithConstField'): withConstFieldNuTauDet = shipGeo.EmuMagnet.WithConstField
       if not withConstFieldNuTauDet:
        fieldMaker.defineFieldMap('NuMap','files/nuTauDetField.root', ROOT.TVector3(0.0,0.0,shipGeo.EmuMagnet.zC))       
+       fieldsList.append('NuMap')    
       withConstFieldHadronAbs = True
       if hasattr(shipGeo.hadronAbsorber,'WithConstField'): withConstFieldHadronAbs = shipGeo.hadronAbsorber.WithConstField
       if not withConstFieldHadronAbs:
        fieldMaker.defineFieldMap('HadronAbsorberMap','files/FieldHadronStopper_raised_20190411.root', ROOT.TVector3(0.0,0.0,shipGeo.hadronAbsorber.z))       
+       fieldsList.append('HadronAbsorberMap')
+      withConstMuonShieldField = True
+      if hasattr (shipGeo, 'muShieldDesign'): 
+       if (shipGeo.muShieldDesign==11): withConstMuonShieldField = False  
+      if not withConstMuonShieldField:
+       zEndOfAbsorb = shipGeo.muShield.z  - shipGeo.muShield.length/2;
+       dZ = [None]*9
+       Z = [None]*9
+       zgap = 10.
+       dZ[0] = shipGeo.muShield.dZ1 - zgap / 2;
+       Z[0] = zEndOfAbsorb + dZ[0] + zgap;
+
+       dZ[1] = shipGeo.muShield.dZ2 - zgap / 2;
+       Z[1] = Z[0] + dZ[0] + dZ[1] + zgap;
+
+       dZ[2] = shipGeo.muShield.dZ3 - zgap / 2;
+       Z[2] = Z[1] + dZ[1] + dZ[2] + 2 * zgap;
+
+       dZ[3] = shipGeo.muShield.dZ4 - zgap / 2;
+       Z[3] = Z[2] + dZ[2] + dZ[3] + zgap;
+
+       dZ[4] = shipGeo.muShield.dZ5 - zgap / 2;
+       Z[4] = Z[3] + dZ[3] + dZ[4] + zgap;
+
+       dZ[5] = shipGeo.muShield.dZ6 - zgap / 2;
+       Z[5] = Z[4] + dZ[4] + dZ[5] + zgap;
+
+       dZ[6] = shipGeo.muShield.dZ7 - zgap / 2;
+       Z[6] = Z[5] + dZ[5] + dZ[6] + zgap;
+
+       dZ[7] = shipGeo.muShield.dZ8 - zgap / 2;
+       Z[7] = Z[6] + dZ[6] + dZ[7] + zgap;
+
+       dZ[8] = 10.;
+       Z[8] = Z[7] + dZ[7] + dZ[8];
+
+       field_center = (Z[2]+Z[8]+dZ[8]-dZ[2])/2
```

Comment:
Looks corret to me. However, have you checked visually that the map coincide with the magnet position. Also, do you know if during field calculation the simulation took into account `zgap` between magents?

Reply:
I have no information about the last part. 
![11](https://user-images.githubusercontent.com/5690500/84252407-35986a00-ab17-11ea-991d-a5508d2435cf.png)
![11y](https://user-images.githubusercontent.com/5690500/84252416-38935a80-ab17-11ea-815f-438387de2148.png)
![Bzx](https://user-images.githubusercontent.com/5690500/84252419-39c48780-ab17-11ea-9706-98fd9a8d20e7.png)
![Bzy](https://user-images.githubusercontent.com/5690500/84252420-39c48780-ab17-11ea-8ff7-29bedf5acc7f.png)

Reply:
BTW, if we are going to use some kind of options object for the shield, we can move this position calculations to the muon shield code.

Reply:
Yeah, this looks valid. Of course, there might be some discrepancies with adding zgap, but I am not aware of the way to easily solve this problem.